spec/models/document_type_selection_spec.rb

```diff
@@ -4,15 +4,22 @@
   let(:document_type_selections) { YAML.load_file(Rails.root.join("config/document_type_selections.yml")) }
 
   describe "all configured document types selections are valid" do
-    let(:document_type_selection_schema) { JSON.parse(File.read("config/schemas/document_type_selection.json")) }
-
     it "conforms to the document type selection schema" do
+      expect(document_type_selections).to all(match_json_schema("document_type_selection"))
+    end
+
+    it "has locale keys that conform to the document type selection locale schema" do
       document_type_selections.each do |document_type_selection|
-        validator = JSON::Validator.fully_validate(document_type_selection_schema, document_type_selection)
-        expect(validator).to(
-          be_empty,
-          "Validation for #{document_type_selection['id']} failed: \n\t#{validator.join("\n\t")}",
-        )
+        translations = I18n.t("document_type_selections.#{document_type_selection['id']}").deep_stringify_keys
+        expect(translations).to match_json_schema("document_type_selection_locale")
+      end
+    end
+
+    it "has locale keys for every option, conforming to the document type selection locale schema" do
+      options = document_type_selections.map { |s| s["options"] }.flatten
+      options.each do |document_type_selection|
+        translations = I18n.t("document_type_selections.#{document_type_selection['id']}").deep_stringify_keys
+        expect(translations).to match_json_schema("document_type_selection_locale")
       end
     end
```

Comment:
It seems a bit strange for us to be testing that these locales match a schema and feels rather unconventional. I seem to remember us previously testing views which felt like a scenario that added more value, what's the rationale for switching and/or are the views problematic?

Reply:
Thanks for the review, Kevin. Agreed it feels unconventional, but so did an attempt at the view tests, since nothing was really _tested_ - we simply visited a bunch of URLs, perhaps making an assertion too as a token gesture. At the time of writing, it would also have meant maintaining a hardcoded array of URLs (now Leena has implemented the `document_type_selection_id` parameter, we could construct a list of all options dynamically).

The schema test does lend some benefits, e.g. failing if an unused property is accidentally added. It makes failure more visible than a bit of text in the console output, and it can be run in isolation as opposed to having to run the whole test suite to test locale usage. Perhaps we should look at this as the beginning of a replacement for the i18n coverage checker in the long term.

In my chat with Ben IRL, we didn't exactly settle on a schema test, but his suggestion was to write a test to grab the locale hash for each document type and check it contains certain keys. This would rewrite a solution already solved by schema tests, IMO, and would also give false positives since the act of retrieving the hash would satisfy the i18n coverage checker even if the hash contained unused keys.

Reply:
I can add a little to this: I explored bumping coverage for fields with view tests, and the result is indeed contrived, since there's nothing to actually test. My worry would be that someone would just delete the tests, since it's not clear what it's testing. The schema (or at least, model-level) approach makes the intent explicit.

For document types, I think the view approach is less clear, since I reckon we'd need to visit all the different branches of the selection tree in order to exercise all the locales.

The schema approach for locales did seem a little strange at first. Since they're just YAML files, it seems reasonable and conventional to use a schema-driven approach to ensure they're valid.